cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/analysis/RangeAnalysisStage.qll

```diff
@@ -920,6 +929,83 @@ module RangeStage<DeltaSig D, BoundSig<D> Bounds, LangSig<D> LangParam, UtilSig<
     bounded(cast.getOperand(), b, delta, upper, fromBackEdge, origdelta, reason)
   }
 
+  predicate bounded(
+    SemExpr e, SemBound b, D::Delta delta, boolean upper, boolean fromBackEdge, D::Delta origdelta,
+    SemReason reason
+  ) {
+    initialBounded(e, b, delta, upper, fromBackEdge, origdelta, reason) and
+    (
+      semExprDoesntOverflow(upper.booleanNot(), e)
+      or
+      not potentiallyOverflowingExpr(upper.booleanNot(), e)
+      or
+      initialBounded(e, any(SemZeroBound z), _, upper.booleanNot(), _, _, _)
+      or
+      exists(D::Delta otherDelta |
+        initialBounded(e, _, otherDelta, upper.booleanNot(), _, _, _) and
+        (
+          upper = true and D::toFloat(otherDelta) >= 0
+          or
+          upper = false and D::toFloat(otherDelta) <= 0
+        )
+      )
```

Comment:
So in order to provide an _upper_ bound (i.e., where `upper = true`) we need to prove that the expression doesn't _underflow_, (because that would provide a very large upper bound that we're not modelling). Is that correct?

And we have four ways of deducing that there's no underflow:
1. Using `semExprDoesntOverflow`, which computes the range of the type of the expression and compares that with whatever absolute bound we can produce.
2. When it's an expression that simply cannot underflow (for example an SSA access).
3. When we have _any_ lower bound computed by `initialBounded`
4. When we have any bound and the delta we compute is non-negative.

Is that a correct summary of the 4 cases above?

A couple of questions:
- Why do we need case 3? Aren't all those cases already included in case 1?
- Does case 4 only apply to relative bounds (since case 3 handles all the non-zero bounds with a looser requirement)